pkg/bloomgateway/worker.go

```diff
@@ -0,0 +1,236 @@
+
+			for _, item := range items {
+				task, ok := item.(Task)
+				if !ok {
+					// This really should never happen, because only the bloom gateway itself can enqueue tasks.
+					return errors.Errorf("failed to cast dequeued item to Task: %v", item)
+				}
+				level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
+				w.tasks.Delete(task.ID)
+
+				fromDay, throughDay := task.Bounds()
+
+				if fromDay.Equal(throughDay) {
+					tasksPerDay[fromDay] = append(tasksPerDay[fromDay], task)
+				} else {
+					for i := fromDay; i.Before(throughDay); i = i.Add(24 * time.Hour) {
+						tasksPerDay[i] = append(tasksPerDay[i], task)
+					}
+				}
+			}
+
+			for day, tasks := range tasksPerDay {
+				logger := log.With(w.logger, "day", day)
+				level.Debug(logger).Log("msg", "process tasks", "tasks", len(tasks))
+
+				storeFetchStart := time.Now()
+				blockRefs, err := w.store.GetBlockRefs(taskCtx, tasks[0].Tenant, day, day.Add(Day).Add(-1*time.Nanosecond))
+				w.metrics.storeAccessLatency.WithLabelValues(w.id, "GetBlockRefs").Observe(time.Since(storeFetchStart).Seconds())
+				if err != nil {
+					for _, t := range tasks {
+						t.ErrCh <- err
+					}
+					// continue with tasks of next day
+					continue
+				}
+				// No blocks found.
+				// Since there are no blocks for the given tasks, we need to return the
+				// unfiltered list of chunk refs.
+				if len(blockRefs) == 0 {
+					level.Warn(logger).Log("msg", "no blocks found")
+					for _, t := range tasks {
+						for _, ref := range t.Request.Refs {
+							t.ResCh <- v1.Output{
+								Fp:       model.Fingerprint(ref.Fingerprint),
+								Removals: nil,
+							}
+						}
+					}
+					// continue with tasks of next day
+					continue
+				}
+
+				boundedRefs := partitionFingerprintRange(tasks, blockRefs)
+				blockRefs = blockRefs[0:]
+				for _, b := range boundedRefs {
+					blockRefs = append(blockRefs, b.blockRef)
+				}
+
+				// GetBlockQueriersForBlockRefs() waits until all blocks are downloaded and available for querying.
+				// TODO(chaudum): Add API that allows to process blocks as soon as they become available.
+				// This will require to change the taskMergeIterator to a slice of requests so we can seek
+				// to the appropriate fingerprint range within the slice that matches the block's fingerprint range.
+				storeFetchStart = time.Now()
+				blockQueriers, err := w.store.GetBlockQueriersForBlockRefs(taskCtx, blockRefs)
+				w.metrics.storeAccessLatency.WithLabelValues(w.id, "GetBlockQueriersForBlockRefs").Observe(time.Since(storeFetchStart).Seconds())
+				if err != nil {
+					for _, t := range tasks {
+						t.ErrCh <- err
+					}
+					// continue with tasks of next day
+					continue
+				}
+
+				for i, blockQuerier := range blockQueriers {
+					it := newTaskMergeIterator(day, boundedRefs[i].tasks...)
+					requests = requests[:0]
+					for it.Next() {
+						requests = append(requests, it.At().Request)
```

Comment:
What's the advantage of collecting these into a slice rather than building an iterator over the underlying iterator?

Reply:
There is none.